src/dune_pkg/toolchain.ml

```diff
@@ -0,0 +1,245 @@
+end
+
+(* A file that gets created in the toolchain directory to indicate
+   that a particular toolchain version is installed. This is what
+   prevents the toolchain from being downloaded/built when it is already
+   installed. *)
+module Done_file = struct
+  let path version = Path.relative (Version.toolchain_dir version) "done"
+  let create version = Path.touch (path version)
+  let exists version = Path.exists (path version)
+end
+
+let handle_checksum_mismatch { Compiler_package.version; url; checksum } ~got_checksum =
+  User_error.raise
+    [ Pp.textf
+        "Checksum mismatch when downloading version %s of the compiler toolchain from %s."
+        (Version.to_string version)
+        (OpamUrl0.to_string url)
+    ; Pp.textf "Expected checksum: %s" (Checksum.to_string checksum)
+    ; Pp.textf "Got checksum: %s" (Checksum.to_string got_checksum)
+    ]
+;;
+
+let handle_unavailable { Compiler_package.version; url; _ } ~msg_opt =
+  let msg_context =
+    Pp.textf
+      "Unable to download version %s of the compiler toolchain from %s."
+      (Version.to_string version)
+      (OpamUrl0.to_string url)
+  in
+  let msg =
+    match (msg_opt : User_message.t option) with
+    | Some msg -> { msg with paragraphs = msg_context :: msg.paragraphs }
+    | None -> User_message.make [ msg_context ]
+  in
+  raise (User_error.E msg)
+;;
+
+let fetch ({ Compiler_package.version; url; checksum } as t) =
+  let open Fiber.O in
+  let source_dir = Version.source_dir version in
+  Fpath.rm_rf (Path.to_string source_dir);
+  Path.mkdir_p source_dir;
+  let+ result =
+    Fetch.fetch
+      ~unpack:true
+      ~checksum:(Some checksum)
+      ~target:source_dir
+      ~url:(Loc.none, url)
+  in
+  match result with
+  | Ok () -> ()
+  | Error (Fetch.Checksum_mismatch got_checksum) ->
+    handle_checksum_mismatch t ~got_checksum
+  | Error (Fetch.Unavailable msg_opt) -> handle_unavailable t ~msg_opt
+;;
+
+let run_command ~dir prog args =
+  let output_on_success =
+    Dune_engine.Execution_parameters.Action_output_on_success.Swallow
+  in
+  let output_limit = Dune_engine.Execution_parameters.Action_output_limit.default in
+  let stdout_to = Process.Io.make_stdout ~output_on_success ~output_limit in
+  let stderr_to = Process.Io.make_stderr ~output_on_success ~output_limit in
+  Process.run ~dir ~stdout_to ~stderr_to ~display:Quiet Strict prog args
+;;
+
+let configure version =
+  let source_dir = Version.source_dir version in
+  let configure_script = Path.relative source_dir "configure" in
+  let prefix = Version.target_dir version in
+  run_command
+    ~dir:source_dir
+    configure_script
+    [ sprintf "--prefix=%s" (Path.to_string prefix) ]
+;;
+
+let make version args =
+  let make = Lazy.force Make.path in
+  let source_dir = Version.source_dir version in
+  run_command
+    ~dir:source_dir
+    make
+    ([ sprintf "--directory=%s" (Version.source_dir version |> Path.to_string) ] @ args)
+;;
+
+let build version = make version [ "-j" ]
```

Comment:
Ah fair. If our realistic options are `-j` or no parallelism at all I say we leave a TODO comment here and keep it as `-j` for now while we're evaluating it. On my M1 mac ocaml takes a bit over a minute to build with `-j` and about 8 minutes to build without so `-j` is much more palatable from a user's point of view.

That said, what are the practical risks of leaving `-j` in without a limit on the number of cores used? Maybe we could test it out on a ubuntu machine under reasonable load, like play a video on youtube while building ocaml with `-j` and see if it causes stuttering or something.

Reply:
Hmm, I think a high number of `-j` could make the build slower (than `$(nproc)`, unlikely slower than just no paralelism), because e.g. of seek times on HDDs and context switching. But yes, a TODO comment can suffice here, especially as this is hopefully not run very often :slightly_smiling_face: